processor/batchprocessor/batch_processor.go

```diff
@@ -41,21 +44,57 @@ import (
 // - cfg.Timeout is elapsed since the timestamp when the previous batch was sent out.
 type batchProcessor struct {
 	logger           *zap.Logger
-	exportCtx        context.Context
-	timer            *time.Timer
 	timeout          time.Duration
 	sendBatchSize    int
 	sendBatchMaxSize int
 
-	newItem chan any
-	batch   batch
+	// batchFunc is a factory for new batch objects corresponding
+	// with the appropriate signal.
+	batchFunc func() batch
+
+	// metadataKeys is the configured list of metadata keys.  When
+	// empty, the `singleton` batcher is used.  When non-empty,
+	// each distinct combination of metadata keys and values
+	// triggers a new batcher, counted in `goroutines`.
+	metadataKeys []string
 
 	shutdownC  chan struct{}
 	goroutines sync.WaitGroup
 
 	telemetry *batchProcessorTelemetry
+
+	// singleton is used when metadataKeys is empty, to avoid the
+	// additional lock and map operations.
+	singleton *batcher
+
+	lock     sync.Mutex
+	batchers map[attribute.Set]*batcher
+}
+
+// batcher is a single instance of the batcher logic.  When metadata
+// keys are in use, one of these is created per distinct combination
+// of values.
+type batcher struct {
+	// processor refers to this processor, for access to common
+	// configuration.
+	processor *batchProcessor
+
+	// exportCtx is a context with the metadata key-values
+	// corresponding with this batcher set.
+	exportCtx context.Context
+
+	// timer informs the batcher send a batch.
+	timer *time.Timer
```

Comment:
Would having a large number of timers (cardinality of the metadata) be a serious problem for the performance?

Reply:
It's a documented feature of this implementation.  I don't think you're asking whether I want to add a limit, and I don't think you're suggesting I should implement a single timer across all the batchers.

I think this approach is good enough to start with, especially since this is a common feature request. If the user implements batching-by-metadata correctly, it will not be a problem. If necessary, the user should install an Auth extension to limit the cardinality of the input--that would be my recommendation if this becomes a problem. 

If pressed, I would implement a hard limit, one where the collector refuses to batch above a hard limit in terms of the cardinality seen so far. WDYT?

Reply:
In my opinion, the added documentation does a good job at explaining the possible risk.

Perhaps what would help is a benchmark of sorts, showing the consumed memory for different cardinalities of metadata. Not sure if it makes sense.

Reply:
It would be good to add a limit that's configurable to guard against unbound number of keys in the metadata. The number of batchers may be useful to track in a metric.

Reply:
Done with `MetadataCardinalityLimit` and an otel metric name `otelcol_batch_cardinality_limit`. Note I've added only an OTel instrument, not an OpenCensus instrument.